File: typescript/src/namespaces/flow.ts
/**
 * Flow namespace — orchestration, routing, approvals, and execution history.
 */

import type { FlowOptions } from '../types';

export class FlowNamespace {
  /** @internal */
  constructor(private callDg: (tool: string, params: Record<string, any>) => Promise<any>,
              private warn: (method: string) => void) {}

  /** Execute a multi-step workflow plan (`data-grout/flow.into`). */
  async run(options: FlowOptions): Promise<any> {
    this.warn('flow.into');
    const params: Record<string, any> = {
      plan: options.plan,
      validate_ctc: options.validateCtc ?? true,
      save_as_skill: options.saveAsSkill ?? false,
    };
    if (options.inputData) params.input_data = options.inputData;
    return this.callDg('flow.into', params);
  }

  /** Conditional dispatch with predicate-based branching (`data-grout/flow.route`). */
  async route(options: {
    branches: Array<Record<string, any>>;
    payload?: any;
    cacheRef?: string;
    else?: any;
    [k: string]: any;
  }): Promise<any> {
    this.warn('flow.route');
    const { branches, payload, cacheRef, else: elseTarget, ...rest } = options;
    const params: Record<string, any> = { branches, ...rest };
    if (payload !== undefined) params.payload = payload;
    if (cacheRef !== undefined) params.cache_ref = cacheRef;
    if (elseTarget !== undefined) params.else = elseTarget;
    return this.callDg('flow.route', params);
  }

  /** Pause workflow for human approval (`data-grout/flow.request-approval`). */
  async requestApproval(options: { action: string; details?: Record<string, any>; reason?: string; context?: Record<string, any>; [k: string]: any }): Promise<any> {
    this.warn('flow.request-approval');
    const { action, details, reason, context, ...rest } = options;
    const params: Record<string, any> = { action, ...rest };
    if (details !== undefined) params.details = details;
    if (reason !== undefined) params.reason = reason;
    if (context !== undefined) params.context = context;
    return this.callDg('flow.request-approval', params);
  }

  /** Request user clarification for missing fields (`data-grout/flow.request-feedback`). */
  async requestFeedback(options: { missing_fields: string[]; reason: string; current_data?: Record<string, any>; suggestions?: Record<string, any>; context?: Record<string, any>; [k: string]: any }): Promise<any> {
    this.warn('flow.request-feedback');
    const { missing_fields, reason, ...rest } = options;
    const params: Record<string, any> = { missing_fields, reason, ...rest };
    return this.callDg('flow.request-feedback', params);
  }

  /** List recent tool executions (`data-grout/inspect.execution-history`). */
  async history(options: { limit?: number; offset?: number; status?: string; refractions_only?: boolean; [k: string]: any } = {}): Promise<any> {
    this.warn('inspect.execution-history');
    const params: Record<string, any> = {
      limit: options.limit ?? 50,
      offset: options.offset ?? 0,
      refractions_only: options.refractions_only ?? false,
      ...options,
    };
    if (options.status !== undefined) params.status = options.status;
    return this.callDg('inspect.execution-history', params);
  }

  /** Get details for a specific execution (`data-grout/inspect.execution-details`). */
  async details(executionId: string): Promise<any> {
    this.warn('inspect.execution-details');
    return this.callDg('inspect.execution-details', { execution_id: executionId });
  }
}

File: rust/src/namespaces/flow.rs
//! Flow namespace — orchestration, routing, approvals, and execution history.

use crate::client::FlowIntoBuilder;
use crate::error::Result;
use crate::Client;
use serde_json::{json, Value};

/// Multi-step orchestration, conditional routing, human-in-the-loop gates,
/// and execution history.
///
/// Obtained via [`Client::flow`].
pub struct Flow<'a>(pub(crate) &'a Client);

impl<'a> Flow<'a> {
    /// Execute a pre-built multi-step workflow plan (`data-grout/flow.into`).
    ///
    /// `plan` is the ordered list of tool-call steps.  Returns a
    /// [`FlowIntoBuilder`] for optional configuration before execution.
    pub fn run(&self, plan: Vec<Value>) -> FlowIntoBuilder<'a> {
        self.0.warn_if_not_dg("flow.into");
        FlowIntoBuilder::new(self.0, plan)
    }

    /// Conditional dispatch with predicate-based branching (`data-grout/flow.route`).
    ///
    /// `branches` is an ordered array of `{ when, then }` objects.  The first
    /// branch whose `when` predicate matches the `payload` (or `cache_ref`
    /// result) fires its `then` target — which can be any tool name, including
    /// another `flow.into` plan or a saved skill.
    ///
    /// # Parameters
    ///
    /// - `branches` (required) — ordered predicate/target pairs
    /// - `payload` — inline data to evaluate predicates against
    /// - `cache_ref` — reference to a cached result to evaluate instead
    /// - `else_target` — fallback tool when no branch matches
    pub async fn route(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("flow.route");
        self.0.ensure_initialized().await?;
        self.0.call_dg_tool("data-grout/flow.route", params).await
    }

    /// Pause workflow for human approval (`data-grout/flow.request-approval`).
    ///
    /// Params: `action`, and optionally `details`, `reason`, `context`.
    pub async fn request_approval(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("flow.request-approval");
        self.0.ensure_initialized().await?;
        self.0
            .call_dg_tool("data-grout/flow.request-approval", params)
            .await
    }

    /// Request user clarification for missing fields (`data-grout/flow.request-feedback`).
    ///
    /// Params: `missing_fields` (array), `reason`, and optionally `current_data`,
    /// `suggestions`, `context`.
    pub async fn request_feedback(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("flow.request-feedback");
        self.0.ensure_initialized().await?;
        self.0
            .call_dg_tool("data-grout/flow.request-feedback", params)
            .await
    }

    /// List recent tool executions (`data-grout/inspect.execution-history`).
    ///
    /// Params: optional `limit`, `offset`, `status`, `refractions_only`.
    pub async fn history(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("inspect.execution-history");
        self.0.ensure_initialized().await?;
        self.0
            .call_dg_tool("data-grout/inspect.execution-history", params)
            .await
    }

    /// Get details for a specific execution (`data-grout/inspect.execution-details`).
    pub async fn details(&self, execution_id: impl Into<String>) -> Result<Value> {
        self.0.warn_if_not_dg("inspect.execution-details");
        self.0.ensure_initialized().await?;
        let params = json!({ "execution_id": execution_id.into() });
        self.0
            .call_dg_tool("data-grout/inspect.execution-details", params)
            .await
    }
}
